crates/mirror-mirror/src/try_visit.rs

```diff
@@ -0,0 +1,204 @@
+    Reflect, ScalarRef,
+};
+use alloc::string::String;
+
+macro_rules! visit_scalar_fn {
+    ($name:ident, $ty:ty) => {
+        #[allow(clippy::ptr_arg)]
+        #[inline]
+        fn $name(&mut self, value: $ty) -> Result<(), Self::Error> {
+            Ok(())
+        }
+    };
+}
+
+#[allow(unused_variables)]
+pub trait TryVisit {
+    type Error;
+
+    visit_scalar_fn!(try_visit_usize, usize);
+    visit_scalar_fn!(try_visit_u8, u8);
+    visit_scalar_fn!(try_visit_u16, u16);
+    visit_scalar_fn!(try_visit_u32, u32);
+    visit_scalar_fn!(try_visit_u64, u64);
+    visit_scalar_fn!(try_visit_u128, u128);
+    visit_scalar_fn!(try_visit_i8, i8);
+    visit_scalar_fn!(try_visit_i16, i16);
+    visit_scalar_fn!(try_visit_i32, i32);
+    visit_scalar_fn!(try_visit_i64, i64);
+    visit_scalar_fn!(try_visit_i128, i128);
+    visit_scalar_fn!(try_visit_bool, bool);
+    visit_scalar_fn!(try_visit_char, char);
+    visit_scalar_fn!(try_visit_f32, f32);
+    visit_scalar_fn!(try_visit_f64, f64);
+    visit_scalar_fn!(try_visit_string, &String);
+
+    #[inline]
+    fn try_visit_opaque(
+        &mut self,
+        value: &dyn Reflect,
+        ty: OpaqueType<'_>,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+pub fn try_visit<V>(visitor: &mut V, value: &dyn Reflect, ty: Type<'_>) -> Result<(), V::Error>
+where
+    V: TryVisit,
+{
+    match ty {
+        Type::Scalar(_) => {
+            let scalar = value.as_scalar().unwrap();
+            match scalar {
+                ScalarRef::usize(inner) => visitor.try_visit_usize(inner)?,
+                ScalarRef::f32(inner) => visitor.try_visit_f32(inner)?,
+                ScalarRef::bool(inner) => visitor.try_visit_bool(inner)?,
+                ScalarRef::u8(inner) => visitor.try_visit_u8(inner)?,
+                ScalarRef::u16(inner) => visitor.try_visit_u16(inner)?,
+                ScalarRef::u32(inner) => visitor.try_visit_u32(inner)?,
+                ScalarRef::u64(inner) => visitor.try_visit_u64(inner)?,
+                ScalarRef::u128(inner) => visitor.try_visit_u128(inner)?,
+                ScalarRef::i8(inner) => visitor.try_visit_i8(inner)?,
+                ScalarRef::i16(inner) => visitor.try_visit_i16(inner)?,
+                ScalarRef::i32(inner) => visitor.try_visit_i32(inner)?,
+                ScalarRef::i64(inner) => visitor.try_visit_i64(inner)?,
+                ScalarRef::i128(inner) => visitor.try_visit_i128(inner)?,
+                ScalarRef::char(inner) => visitor.try_visit_char(inner)?,
+                ScalarRef::f64(inner) => visitor.try_visit_f64(inner)?,
+                ScalarRef::String(inner) => visitor.try_visit_string(inner)?,
+            }
+        }
+        Type::Struct(struct_ty) => {
+            let struct_ = value.as_struct().unwrap();
+
+            for field_ty in struct_ty.field_types() {
+                let field = struct_.field(field_ty.name()).unwrap();
+                try_visit(visitor, field, field_ty.get_type())?;
```

Comment:
Do you need a stop condition (or a set of visited nodes) to handle recursive data types?
Independently of the answer, can you add a test that has a `TryVisit` implementation traversing a recursive data type, please?

Reply:
No I don't think so. Since we recurse based on the values, and not the types, we'll only recurse as deeply as there are values. Not sure if that makes sense 😅 but I've added a test that at least demonstrates that.

Reply:
It does, the memory can't hold an infinite amount of values. Thanks for adding a test :+1:

Reply:
I assume that you'll run into other problems first if you have such a large value 😜